// @microsoft/power-apps-cli/dist/HttpClient/CliHttpClient.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { randomUUID } from 'crypto';
import { getSettings } from '../CliSettings.js';
import { MS_CLIENT_REQUESTID } from '../Constants.js';
import { enhanceNetworkError } from '../Utils/EnhanceNetworkError.js';
import { CliSpinner } from '../Utils/Spinner.js';
export class CliHttpClient {
    _authenticationProvider;
    _inflight = 0;
    _spinner = new CliSpinner();
    _firstAuthCallbacks = [];
    _firstAuthFired = false;
    constructor(_authenticationProvider) {
        this._authenticationProvider = _authenticationProvider;
    }
    /**
     * Subscribes a callback to fire once, in the background, after the first successful authenticated
     * request (i.e. the first time a token is actually acquired). Multiple subscribers are supported —
     * all registered callbacks fire on that first auth. Used by `CliLogger` to upgrade telemetry to
     * the authed writer only once auth is known to work, so telemetry can never be the thing that
     * triggers an interactive sign-in.
     */
    onFirstAuth(callback) {
        this._firstAuthCallbacks.push(callback);
    }
    async get(url, config) {
        return this._sendRequestAsync(url, 'GET', config);
    }
    async post(url, config) {
        return this._sendRequestAsync(url, 'POST', config);
    }
    async patch(url, config) {
        return this._sendRequestAsync(url, 'PATCH', config);
    }
    async put(url, config) {
        return this._sendRequestAsync(url, 'PUT', config);
    }
    async delete(url, config) {
        return this._sendRequestAsync(url, 'DELETE', config);
    }
    async _sendRequestAsync(url, method, config) {
        this._startSpinner();
        try {
            const headers = config?.headers || {};
            if (config?.authResource) {
                const token = await this._authenticationProvider.getAccessTokenForResource(config.authResource);
                headers.Authorization = `Bearer ${token}`;
                // Successful token acquisition is the lazy-auth signal — fire the registered callbacks once,
                // in the background, so they don't block this request.
                if (token && !this._firstAuthFired && this._firstAuthCallbacks.length > 0) {
                    this._firstAuthFired = true;
                    for (const callback of this._firstAuthCallbacks) {
                        callback().catch(() => {
                            // Each callback owns its own error handling; a background failure must not affect the
                            // outcome of the current request.
                        });
                    }
                }
            }
            if (!headers['Content-Type']) {
                headers['Content-Type'] = 'application/json';
            }
            headers[MS_CLIENT_REQUESTID] = randomUUID();
            // Only JSON.stringify if body is not already a string and Content-Type is JSON
            let bodyContent;
            if (config?.body !== undefined) {
                const isJsonContentType = headers['Content-Type'] === 'application/json';
                bodyContent =
                    typeof config.body === 'string' && !isJsonContentType
                        ? config.body
                        : JSON.stringify(config.body);
            }
            const requestOptions = {
                method,
                headers,
                // Only include body if it exists
                body: bodyContent,
            };
            const response = await fetch(url, requestOptions);
            if (!response.ok) {
                throw new Error(`HTTP error status: ${response.status} for ${method} ${url}: ${await response.text()}`);
            }
            const data = await this._parseResponse(response);
            return {
                data,
                status: response.status,
                headers: await this._transformHeaders(response.headers),
            };
        }
        catch (error) {
            throw enhanceNetworkError(error, method, url);
        }
        finally {
            this._stopSpinner();
        }
    }
    _startSpinner() {
        this._inflight++;
        if (this._inflight === 1 && getSettings().interactive) {
            this._spinner.start();
        }
    }
    _stopSpinner() {
        this._inflight--;
        if (this._inflight === 0 && getSettings().interactive) {
            this._spinner.stop();
        }
    }
    async _parseResponse(response) {
        const contentType = response.headers.get('content-type');
        const responseText = await response.text();
        // Handle empty responses or non-JSON content
        if (!responseText || responseText.trim() === '') {
            return null;
        }
        else if (contentType && contentType.includes('application/json')) {
            try {
                return JSON.parse(responseText);
            }
            catch (parseError) {
                throw new Error(`Failed to parse JSON response: ${parseError}`);
            }
        }
        else {
            return responseText;
        }
    }
    async _transformHeaders(headers) {
        const result = {};
        headers.forEach((value, key) => {
            result[key] = value;
        });
        return result;
    }
}
