// @microsoft/power-apps-cli/dist/CliUtils.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import * as prompts from '@clack/prompts';
import { isCancel } from '@clack/prompts';
import { getPlayerBaseUrl, PowerPlatformApiDiscovery, } from '@microsoft/power-apps-common/services';
import { printWarning } from './Utils/ConsoleOutput.js';
/**
 * Prompts the user for text input. Exits the process if the user cancels.
 */
export async function promptText(options) {
    const result = await prompts.text(options);
    if (isCancel(result)) {
        printWarning('Operation cancelled by user.');
        process.exit(0);
    }
    return result;
}
/**
 * Prompts the user for confirmation. Exits the process if the user cancels.
 */
export async function promptConfirm(options) {
    const result = await prompts.confirm(options);
    if (isCancel(result)) {
        printWarning('Operation cancelled by user.');
        process.exit(0);
    }
    return result;
}
/**
 * Prompts the user to select from a list of options. Exits the process if the user cancels.
 */
export async function promptSelect(options) {
    // clack 0.6's select option typing doesn't simplify for a generic `T extends string`, so we
    // cast to satisfy the constraint at the call site only. The runtime contract is unchanged.
    const result = await prompts.select(options);
    if (isCancel(result)) {
        printWarning('Operation cancelled by user.');
        process.exit(0);
    }
    return result;
}
/**
 * Token resource (audience) for the Power Platform API in the given region — the resource the
 * `login` verb requests an access token for. `local`/`dev` fall back to the `test` audience
 * (there is no dedicated local token audience). Mirrors `getAuthResource()` in
 * `@microsoft/power-apps-actions`, kept here so the CLI doesn't depend on a non-exported helper.
 */
export function getAuthResourceUrl(region) {
    switch (region) {
        case 'preprod':
            return new PowerPlatformApiDiscovery('preprod').getTokenAudience();
        case 'preview':
            return new PowerPlatformApiDiscovery('firstrelease').getTokenAudience();
        case 'prod':
            return new PowerPlatformApiDiscovery('prod').getTokenAudience();
        case 'gccmoderate':
            return new PowerPlatformApiDiscovery('gov').getTokenAudience();
        case 'gcchigh':
            return new PowerPlatformApiDiscovery('high').getTokenAudience();
        case 'dod':
            return new PowerPlatformApiDiscovery('dod').getTokenAudience();
        case 'mooncake':
            return new PowerPlatformApiDiscovery('mooncake').getTokenAudience();
        case 'ex':
            return new PowerPlatformApiDiscovery('ex').getTokenAudience();
        case 'rx':
            return new PowerPlatformApiDiscovery('rx').getTokenAudience();
        case 'test':
        case 'local':
        case 'dev':
        default:
            return new PowerPlatformApiDiscovery('test').getTokenAudience();
    }
}
export function createDefaultStringOption(options) {
    const { flags, env, message, initialValue, description, default: defaultValue } = options;
    return {
        flags,
        env,
        description,
        default: defaultValue,
        customPrompt: async () => {
            return await promptText({ message, initialValue });
        },
    };
}
export function isGuid(value) {
    const guidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    return guidRegex.test(value);
}
export function createPlayUrl(region, environmentId, appId, localAppUrl, localConnectionsUrl) {
    const baseUrl = getPlayerBaseUrl(region);
    const playUrl = `${baseUrl}/play/e/${environmentId}/app/${appId}`;
    const queryParams = [];
    if (localAppUrl) {
        queryParams.push(`_localAppUrl=${localAppUrl}`);
    }
    if (localConnectionsUrl) {
        queryParams.push(`_localConnectionUrl=${localConnectionsUrl}`);
    }
    return queryParams.length > 0 ? `${playUrl}?${queryParams.join('&')}` : playUrl;
}
export function getAuthority(region, tenantId) {
    switch (region) {
        case 'gcchigh':
        case 'dod':
            return `https://login.microsoftonline.us/${tenantId || 'organizations'}`;
        case 'mooncake':
            return `https://login.partner.microsoftonline.cn/${tenantId || 'organizations'}`;
        default:
            return `https://login.microsoftonline.com/${tenantId || 'organizations'}`;
    }
}
