// @microsoft/power-apps-cli/dist/Verbs/AuthStatus.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { HelpStrings } from '../Constants/HelpStrings.js';
import { printInfo, printTable, printWarning } from '../Utils/ConsoleOutput.js';
export async function authStatusAsync(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Auth-status'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['Auth-status'].Examples ?? []);
    await argumentProvider.runVerb();
    const accounts = await argumentProvider.getAccountManager().listCachedAccountsAsync();
    const jsonMode = argumentProvider.isJsonMode();
    if (jsonMode) {
        const active = accounts.find((a) => a.isActive);
        console.log(JSON.stringify({
            signedIn: accounts.length > 0,
            activeAccount: active
                ? { username: active.username, homeAccountId: active.homeAccountId }
                : null,
            accounts: accounts.map((a) => ({
                username: a.username,
                homeAccountId: a.homeAccountId,
                isActive: a.isActive,
            })),
        }, null, 2));
        return;
    }
    if (accounts.length === 0) {
        printWarning('Not signed in. Run `npx power-apps login` to authenticate.');
        return;
    }
    if (accounts.length === 1) {
        printInfo(`Signed in as ${accounts[0].username}`);
        return;
    }
    const hasActive = accounts.some((a) => a.isActive);
    printTable({
        title: 'Cached accounts',
        columns: [{ header: 'Active' }, { header: 'Username' }],
        rows: accounts.map((a) => [a.isActive ? '*' : '', a.username]),
    });
    if (!hasActive) {
        printWarning('No active account selected. Run `npx power-apps auth-switch` to choose one.');
    }
}
