// @microsoft/power-apps-cli/dist/Verbs/AuthSwitch.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getSettings } from '../CliSettings.js';
import { promptSelect } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { UsageError } from '../Errors/CliError.js';
import { printSuccess, printWarning } from '../Utils/ConsoleOutput.js';
import { AccountArguments } from './VerbConstants.js';
export async function authSwitchAsync(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Auth-switch'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['Auth-switch'].Examples ?? []);
    argumentProvider.addOption({
        flags: { key: AccountArguments.Account.name },
        env: AccountArguments.Account.envVar,
        description: HelpStrings.Verbs['Auth-switch'].Options?.Account ?? '',
        optionalInNonInteractive: true,
    });
    await argumentProvider.runVerb();
    const accountManager = argumentProvider.getAccountManager();
    const requested = await argumentProvider.getOption(AccountArguments.Account.name);
    const jsonMode = argumentProvider.isJsonMode();
    const accounts = await accountManager.listCachedAccountsAsync();
    if (accounts.length === 0) {
        if (jsonMode) {
            console.log(JSON.stringify({ activeAccount: null }, null, 2));
            return;
        }
        printWarning('No cached accounts. Run `npx power-apps login` first.');
        return;
    }
    let target = requested;
    if (!target) {
        if (accounts.length === 1) {
            // No-op confirmation: the only account is also the active one.
            const onlyAccount = await accountManager.setActiveAccountAsync(accounts[0].homeAccountId);
            _emitSuccess(onlyAccount.username, onlyAccount.homeAccountId, jsonMode);
            return;
        }
        // JSON-mode callers can't sensibly answer an interactive picker, so we treat `--json` as
        // implying non-interactive for the disambiguation prompt — same way we treat
        // `--non-interactive` itself.
        if (jsonMode || !getSettings().interactive) {
            throw new UsageError(`Multiple accounts are cached. Pass --account <email> to choose one.\n` +
                accounts.map((a) => `  - ${a.username}`).join('\n'));
        }
        target = await promptSelect({
            message: 'Choose the active account:',
            options: accounts.map((a) => ({
                value: a.homeAccountId,
                label: a.username,
                hint: a.isActive ? 'currently active' : undefined,
            })),
        });
    }
    const resolved = await accountManager.setActiveAccountAsync(target);
    _emitSuccess(resolved.username, resolved.homeAccountId, jsonMode);
}
function _emitSuccess(username, homeAccountId, jsonMode) {
    if (jsonMode) {
        console.log(JSON.stringify({ activeAccount: { username, homeAccountId } }, null, 2));
    }
    else {
        printSuccess(`Active account is ${username}.`);
    }
}
