// @microsoft/power-apps-cli/dist/Verbs/CreateConnection.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { createConnectionAsync as createConnectionAction, getConnectorAsync, getPlayerServiceConfig, isSsoOnlyConnector, } from '@microsoft/power-apps-actions';
import { createMaafConnectionUrl } from '@microsoft/power-apps-common/services';
import * as http from 'http';
import open from 'open';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { CliError, UsageError } from '../Errors/CliError.js';
import { printInfo, printSuccess, printWarning } from '../Utils/ConsoleOutput.js';
import { CreateConnectionArguments, SharedArguments } from './VerbConstants.js';
const CONNECTION_CREATION_TIMEOUT_MS = 10 * 60 * 1000; // 10 minutes
const PROTOCOL_VERSION = '1';
export const BROWSER_CONNECTION_ENV_VAR = 'POWERAPPS_CLI_ENABLE_BROWSER_CONNECTION';
export const BROWSER_DISABLED_MESSAGE = 'This connection requires user interaction. Please open https://make.powerapps.com to create a connection.';
export function isBrowserConnectionEnabled() {
    const value = process.env[BROWSER_CONNECTION_ENV_VAR];
    if (!value) {
        return false;
    }
    const normalized = value.trim().toLowerCase();
    return normalized === '1' || normalized === 'true' || normalized === 'yes' || normalized === 'on';
}
/**
 * Create a connection for a connector.
 *
 * Behavior:
 *  - Fetches the connector definition.
 *  - If the connector has a single SSO-eligible auth type, attempts silent SSO
 *    via `createConnectionAsync` (no browser required).
 *  - Otherwise (or if SSO fails in interactive mode), opens the maaf-player
 *    connection-creation page in the browser and waits for the callback.
 *  - In non-interactive mode without SSO-eligibility, fails with a clear
 *    `UsageError` since there is no browser to fall back to.
 */
export async function createConnectionVerbAsync(argumentProvider) {
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const settings = getSettings();
    const logger = getCliLogger();
    const playerServiceConfig = getPlayerServiceConfig();
    const authProvider = argumentProvider.getAuthenticationProvider();
    const apiIdRaw = await argumentProvider.getOption('apiId');
    const displayName = await argumentProvider.getOption('displayName', true);
    if (!apiIdRaw) {
        throw new UsageError(`Missing required option --${SharedArguments.ApiId.name} (-${SharedArguments.ApiId.alias}).`);
    }
    const apiId = normalizeConnectorId(apiIdRaw);
    // 1. Fetch connector + decide SSO vs browser path
    const connector = await getConnectorAsync(apiId, logger);
    if (!connector) {
        throw new CliError(`Connector '${apiId}' not found.`);
    }
    // SSO-only path: try silent flow
    if (isSsoOnlyConnector(connector)) {
        try {
            const context = {
                vfs: settings.cliFs,
                authProvider,
                region: playerServiceConfig.region,
                environmentName: playerServiceConfig.environmentName,
                actionsParams: { connectorId: apiId, displayName },
                localFilePaths: {
                    powerConfigPath: settings.fileConfig.powerConfigPath,
                    schemaPath: settings.fileConfig.schemaPath,
                    codeGenPath: settings.fileConfig.codeGenPath,
                },
                logger,
            };
            const connection = await createConnectionAction(context);
            emitSuccess(argumentProvider, {
                connectionId: connection.name,
                displayName: connection.properties.displayName,
            });
            return;
        }
        catch (ssoError) {
            if (!settings.interactive) {
                throw ssoError;
            }
            if (!isBrowserConnectionEnabled()) {
                printWarning(`Silent connection creation failed: ${ssoError instanceof Error ? ssoError.message : String(ssoError)}`);
                throw new UsageError(BROWSER_DISABLED_MESSAGE);
            }
            printWarning(`Silent connection creation failed: ${ssoError instanceof Error ? ssoError.message : String(ssoError)}`);
            printInfo('Falling back to browser-based connection creation...\n');
        }
    }
    else if (!settings.interactive) {
        throw new UsageError(`Connector '${apiId}' requires interactive sign-in to create a connection. ` +
            `Re-run without --non-interactive, or create the connection in the Power Apps maker portal.`);
    }
    else if (!isBrowserConnectionEnabled()) {
        throw new UsageError(BROWSER_DISABLED_MESSAGE);
    }
    else {
        printInfo('This connector requires manual interaction. Opening browser for connection creation...\n');
    }
    // 2. Browser-based fallback
    const result = await startConnectionServerAsync({
        connectorName: apiId,
        environmentId: playerServiceConfig.environmentName,
        region: playerServiceConfig.region,
    });
    if (result.status === 'created' && result.connectionName) {
        emitSuccess(argumentProvider, {
            connectionId: result.connectionName,
            displayName: result.displayName,
        });
    }
    else if (result.status === 'error') {
        throw new CliError(result.errorMessage ?? 'Connection creation failed in the browser.');
    }
    else {
        throw new UsageError('Connection creation was cancelled.');
    }
}
function emitSuccess(argumentProvider, result) {
    if (argumentProvider.isJsonMode()) {
        console.log(JSON.stringify(result, null, 2));
        return;
    }
    printSuccess('Connection created successfully!');
    if (result.displayName) {
        printInfo(`  Display Name: ${result.displayName}`);
    }
    printInfo(`  Connection ID: ${result.connectionId}`);
}
function normalizeConnectorId(connectorId) {
    return connectorId.startsWith('shared_') ? connectorId : `shared_${connectorId}`;
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Create-connection'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['Create-connection'].Examples ?? []);
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: SharedArguments.ApiId.name,
                alias: SharedArguments.ApiId.alias,
            },
            env: SharedArguments.ApiId.envVar,
            message: SharedArguments.ApiId.promptMessage,
            description: HelpStrings.Verbs['Create-connection'].Options?.ApiId ?? '',
        }),
        createDefaultStringOption({
            flags: {
                key: CreateConnectionArguments.DisplayName.name,
                alias: CreateConnectionArguments.DisplayName.alias,
            },
            env: CreateConnectionArguments.DisplayName.envVar,
            message: CreateConnectionArguments.DisplayName.promptMessage,
            description: HelpStrings.Verbs['Create-connection'].Options?.DisplayName ?? '',
        }),
    ]);
}
/**
 * Starts a local HTTP server, opens the maaf-player connection page in the
 * browser, and waits for a callback with the result. Uses the same pattern as
 * an OAuth authorization code flow: nonce for CSRF, protocol version for drift
 * detection, and a timeout for cleanup.
 */
async function startConnectionServerAsync(config) {
    const { connectorName, environmentId, region } = config;
    const nonce = crypto.randomUUID();
    return new Promise((resolve, reject) => {
        const server = http.createServer((req, res) => {
            const url = new URL(req.url ?? '/', `http://localhost`);
            if (url.pathname === '/callback' && req.method === 'GET') {
                const queryNonce = url.searchParams.get('nonce');
                if (queryNonce !== nonce) {
                    res.writeHead(403, { 'Content-Type': 'text/html; charset=utf-8' });
                    res.end('<html><body><h2>Connection request verification failed. Please run the CLI command again.</h2></body></html>');
                    cleanup();
                    resolve({ status: 'cancelled' });
                    return;
                }
                const callbackVersion = url.searchParams.get('v');
                if (callbackVersion !== PROTOCOL_VERSION) {
                    printWarning(`Warning: callback protocol version mismatch (expected ${PROTOCOL_VERSION}, got ${callbackVersion ?? 'none'}). Update the CLI to the latest version.`);
                }
                const status = url.searchParams.get('status');
                const errorMessage = url.searchParams.get('message') ?? undefined;
                res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
                let tabHtml;
                if (status === 'created') {
                    tabHtml =
                        '<html><body><h2>Connection created! You can close this tab.</h2></body></html>';
                }
                else if (status === 'error') {
                    const safeMessage = errorMessage
                        ? errorMessage.replace(/[&<>"']/g, (c) => `&#${c.charCodeAt(0)};`)
                        : '';
                    tabHtml = safeMessage
                        ? `<html><body><h2>Connection creation failed. You can close this tab.</h2><p>${safeMessage}</p></body></html>`
                        : '<html><body><h2>Connection creation failed. You can close this tab.</h2></body></html>';
                }
                else {
                    tabHtml =
                        '<html><body><h2>Connection creation cancelled. You can close this tab.</h2></body></html>';
                }
                res.end(tabHtml);
                cleanup();
                if (status === 'created') {
                    resolve({
                        status: 'created',
                        connectionId: url.searchParams.get('connectionId') ?? undefined,
                        connectionName: url.searchParams.get('connectionName') ?? undefined,
                        displayName: url.searchParams.get('displayName') ?? undefined,
                    });
                }
                else if (status === 'error') {
                    resolve({ status: 'error', errorMessage });
                }
                else {
                    resolve({ status: 'cancelled' });
                }
                return;
            }
            res.writeHead(404);
            res.end('Not found');
        });
        server.listen(0, 'localhost', async () => {
            const address = server.address();
            const port = typeof address === 'object' && address ? address.port : 0;
            const callbackUrl = `http://localhost:${port}/callback`;
            const playerUrl = createMaafConnectionUrl(region, environmentId, {
                connector: connectorName,
                callbackUrl,
                nonce,
                protocolVersion: PROTOCOL_VERSION,
            });
            printInfo(`Opening browser at ${playerUrl}`);
            try {
                await open(playerUrl, { wait: false });
            }
            catch {
                printInfo(`Could not open browser automatically. Please navigate to:\n  ${playerUrl}`);
            }
            printInfo('Waiting for connection creation... (press Ctrl+C to cancel)\n');
        });
        server.on('error', (error) => {
            cleanup();
            reject(error);
        });
        const timeoutId = setTimeout(() => {
            printInfo('\nConnection creation timed out after 10 minutes.');
            cleanup();
            resolve({ status: 'cancelled' });
        }, CONNECTION_CREATION_TIMEOUT_MS);
        const sigintHandler = () => {
            printInfo('\nCancelling...');
            cleanup();
            resolve({ status: 'cancelled' });
        };
        function cleanup() {
            clearTimeout(timeoutId);
            process.off('SIGINT', sigintHandler);
            server.close();
        }
        process.on('SIGINT', sigintHandler);
    });
}
