// @microsoft/power-apps-cli/dist/Authentication/NodeMsalAuthenticationProvider.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { InteractionRequiredAuthError, PublicClientApplication } from '@azure/msal-node';
import open from 'open';
import { getSettings } from '../CliSettings.js';
import { getAuthority } from '../CliUtils.js';
import { AUTH_CACHE_DIRECTORY, CLI_CLIENT_ID } from '../Constants.js';
import { UsageError } from '../Errors/CliError.js';
import { printError, printInfo, printWarning } from '../Utils/ConsoleOutput.js';
import { ActiveAccountStore } from './ActiveAccountStore.js';
import { clearAuthCache } from './clearAuthCache.js';
/**
 * Interactive MSAL provider. Implements both the token surface (`IAuthenticationProvider`)
 * and the account-management surface (`IAccountManager`). Built on `@azure/msal-node`'s
 * `PublicClientApplication` with the `msal-node-extensions` DPAPI-encrypted persistent cache.
 */
export class NodeMsalAuthenticationProvider {
    _msalClient;
    _tenantId;
    _region;
    _activeStore;
    _cacheDirectory;
    // Resolves whether the CLI is emitting JSON — used only to suppress the interactive
    // sign-in status line so it can't corrupt a `--json` stdout payload. Defaults to a conservative
    // argv peek for direct/test construction; `Cli.ts` wires the authoritative
    // `ArgumentProvider.isJsonMode()` via `setJsonModeResolver()` once that exists.
    _isJsonMode = () => process.argv.includes('--json');
    /**
     * `cacheDirectory` defaults to the production auth cache (`~/.powerapps-cli/cache/auth/`).
     * Tests pass a tmpdir so they can verify cache-write/clear behavior without touching the
     * user's real cache.
     */
    constructor(cacheDirectory = AUTH_CACHE_DIRECTORY) {
        this._cacheDirectory = cacheDirectory;
        this._activeStore = new ActiveAccountStore(cacheDirectory);
    }
    async initAsync(region) {
        this._region = region;
        const { DataProtectionScope, PersistenceCachePlugin, PersistenceCreator } = await import('@azure/msal-node-extensions');
        const persistenceConfiguration = {
            cachePath: this._cacheDirectory + '/msal_cache.json',
            dataProtectionScope: DataProtectionScope.CurrentUser,
            serviceName: 'power-apps',
            accountName: 'power-apps',
            usePlaintextFileOnLinux: false,
        };
        const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration);
        const authConfig = {
            auth: {
                authority: getAuthority(this._region, this._tenantId),
                clientId: CLI_CLIENT_ID,
            },
            cache: {
                cachePlugin: new PersistenceCachePlugin(persistence),
            },
        };
        this._msalClient = new PublicClientApplication(authConfig);
    }
    /**
     * Injects the CLI's authoritative JSON-mode check (`ArgumentProvider.isJsonMode`), used only to
     * suppress the interactive sign-in status line so it never corrupts a `--json` stdout payload.
     * Wired from `Cli.ts` once the `ArgumentProvider` exists; until then the conservative argv-peek
     * default is used (e.g. unit tests that construct the provider directly).
     */
    setJsonModeResolver(isJsonMode) {
        this._isJsonMode = isJsonMode;
    }
    // ─── IAuthenticationProvider ─────────────────────────────────────────────────
    async getAccessTokenForResource(resource, loginHint) {
        const result = await this._acquireTokenInternal(resource + '/.default', loginHint);
        this._tenantId = result.tenantId;
        return result.accessToken;
    }
    getUserTenantId() {
        return this._tenantId;
    }
    /**
     * Returns the active cached account's username, or `undefined` when not signed in (zero accounts,
     * or 2+ accounts with no active pointer). Pure cache read via `_getAccounts` + `_pickActiveAccount`
     * — never acquires a token or opens a browser, so telemetry can call it safely at startup.
     */
    async getCachedAccountAsync() {
        const accounts = await this._getAccounts();
        return this._pickActiveAccount(accounts)?.username;
    }
    // ─── IAccountManager ─────────────────────────────────────────────────────────
    async listCachedAccountsAsync() {
        const accounts = await this._getAccounts();
        if (accounts.length === 0)
            return [];
        const active = this._pickActiveAccount(accounts);
        return accounts.map((account) => ({
            homeAccountId: account.homeAccountId,
            username: account.username,
            isActive: account.homeAccountId === active?.homeAccountId,
        }));
    }
    async setActiveAccountAsync(accountKey) {
        const accounts = await this._getAccounts();
        const match = this._findAccount(accounts, accountKey);
        if (!match) {
            throw new UsageError(`No cached account matches '${accountKey}'. Run \`npx power-apps auth-status\` to list cached accounts.`);
        }
        // Idempotent (mirrors `_ensureActivePointer`): only rewrite the pointer when it isn't already
        // this account, so a no-op `auth-switch` on an already-active account doesn't touch the file.
        const existing = this._activeStore.read();
        if (existing?.homeAccountId !== match.homeAccountId) {
            this._activeStore.write({ homeAccountId: match.homeAccountId, username: match.username });
        }
        return { homeAccountId: match.homeAccountId, username: match.username, isActive: true };
    }
    async loginInteractiveAsync(resource, loginHint) {
        // Two paths, mutually exclusive:
        //   1. `--account <email>` was passed → use the value as MSAL `loginHint` and DON'T force the
        //      picker. Entra honors the hint (SSO if a session matches, otherwise pre-fills the form).
        //   2. No `--account` → force `prompt: 'select_account'` so the user can pick an existing
        //      account or add a new one. Without this, the browser's SSO session would silently reuse
        //      the same account on every `login` — defeating the multi-account scenario.
        // Mixing `loginHint` with `prompt: 'select_account'` gives the picker UI even when the user
        // asked for a specific account, so we pass one or the other but never both.
        const result = await this._signInInteractiveAndPersist(resource + '/.default', loginHint, {
            prompt: loginHint ? undefined : 'select_account',
        });
        if (!result.account) {
            throw new Error('Interactive sign-in did not return an account.');
        }
        return {
            homeAccountId: result.account.homeAccountId,
            username: result.account.username,
            isActive: true,
        };
    }
    async clearCacheAsync() {
        return clearAuthCache(this._cacheDirectory);
    }
    // ─── Internals ───────────────────────────────────────────────────────────────
    async _getAccounts() {
        if (!this._msalClient) {
            throw new Error('Authentication not initialized yet');
        }
        try {
            return await this._msalClient.getTokenCache().getAllAccounts();
        }
        catch (err) {
            // Most likely cause is a cache file written by an older CLI version that the current MSAL
            // major can't parse. Treat as "no cached accounts" so the next command triggers a fresh
            // interactive sign-in, which rewrites the cache in the current format.
            printWarning(`Could not read the MSAL token cache (${err.message}). ` +
                `It may have been written by an older CLI version. Run \`npx power-apps logout\` and sign in again to reset it.`);
            return [];
        }
    }
    /**
     * Case-insensitive lookup against `homeAccountId` first, then `username`. The two-field search
     * lets callers pass either form without a discriminator. Username has a known collision shape:
     * the same UPN can appear in multiple tenants for guest accounts; when the caller might hit
     * that, prefer `homeAccountId`.
     */
    _findAccount(accounts, key) {
        const lower = key.toLowerCase();
        return (accounts.find((a) => a.homeAccountId.toLowerCase() === lower) ??
            accounts.find((a) => a.username.toLowerCase() === lower));
    }
    /**
     * Pure. Resolves which cached account should be treated as active:
     *   0 accounts                           → undefined
     *   1 account                            → that one (regardless of pointer state)
     *   2+ + valid active pointer            → the matching account
     *   2+ + missing or stale active pointer → undefined (caller decides how to handle)
     */
    _pickActiveAccount(accounts) {
        if (accounts.length === 0)
            return undefined;
        if (accounts.length === 1)
            return accounts[0];
        const active = this._activeStore.read();
        if (!active)
            return undefined;
        return accounts.find((a) => a.homeAccountId === active.homeAccountId);
    }
    /**
     * Idempotent: writes the active-account pointer only when the persisted value doesn't already
     * match. Called from token-acquisition success paths so that an upgrade from the old
     * single-account CLI (one account in cache, no pointer file yet) migrates implicitly on first
     * use, without needing an explicit migration step.
     */
    _ensureActivePointer(account) {
        const existing = this._activeStore.read();
        if (existing?.homeAccountId === account.homeAccountId)
            return;
        this._activeStore.write({
            homeAccountId: account.homeAccountId,
            username: account.username,
        });
    }
    async _acquireTokenInternal(resource, loginHint) {
        if (!this._msalClient) {
            throw new Error('Authentication not initialized yet');
        }
        const accounts = await this._getAccounts();
        if (accounts.length === 0) {
            this._ensureInteractiveSignInAllowed('no-account');
            return this._signInInteractiveAndPersist(resource, loginHint);
        }
        const active = this._pickActiveAccount(accounts);
        if (!active) {
            // 2+ accounts and no valid active pointer — refuse rather than guess.
            const names = accounts.map((a) => `  - ${a.username}`).join('\n');
            const hint = getSettings().interactive
                ? `Run \`npx power-apps auth-switch\` to choose one, or \`npx power-apps logout\` to clear all cached accounts and start over.`
                : `Run \`npx power-apps auth-switch --account <email>\` to resolve, then retry.`;
            throw new UsageError(`Multiple accounts are cached and no active account is selected:\n${names}\n${hint}`);
        }
        try {
            const result = await this._msalClient.acquireTokenSilent({
                account: active,
                scopes: [resource],
            });
            this._ensureActivePointer(result.account ?? active);
            return result;
        }
        catch (error) {
            if (error instanceof InteractionRequiredAuthError) {
                this._ensureInteractiveSignInAllowed('expired');
                return this._signInInteractiveAndPersist(resource, loginHint ?? active.username);
            }
            throw error;
        }
    }
    /**
     * Guards the interactive sign-in entry points. Interactive sign-in opens a browser, which must
     * never happen in non-interactive mode (CI, piped output, `--non-interactive`) — there the
     * silent paths (zero cached accounts, or an expired/invalid refresh token that MSAL reports as
     * `InteractionRequiredAuthError`) must fail fast with an actionable error rather than hijack the
     * terminal with a browser window that no one is watching.
     */
    _ensureInteractiveSignInAllowed(reason) {
        if (getSettings().interactive)
            return;
        const cause = reason === 'no-account'
            ? 'You are not signed in'
            : 'Your sign-in session has expired and could not be renewed silently';
        throw new UsageError(`${cause}. Interactive sign-in is required, but the CLI is running in non-interactive mode. ` +
            `Run \`npx power-apps login\` to sign in, then retry.`);
    }
    /**
     * Single code path for every interactive sign-in: `login`, the 0-accounts fresh-login
     * branch, and the `InteractionRequiredAuthError` fallback all route through here. Keeps the
     * "open browser → record tenant → persist active pointer" sequence in one place. Uses the
     * idempotent `_ensureActivePointer` so an unchanged-active scenario doesn't rewrite the pointer.
     *
     * `options.prompt` forwards to MSAL's `prompt` parameter. Callers opt in only when they need to
     * override the IdP's default — `login` passes `'select_account'` so users can add a second
     * account; the silent-fallback paths leave it unset so token expiry doesn't pop a picker.
     */
    async _signInInteractiveAndPersist(resource, loginHint, options) {
        if (!this._msalClient) {
            throw new Error('Authentication not initialized yet');
        }
        const request = this._getInteractiveLoginRequest(resource, loginHint);
        if (options?.prompt) {
            request.prompt = options.prompt;
        }
        // `acquireTokenInteractive` blocks until the browser round-trip completes, which can take many
        // seconds while the user signs in. Without an indicator the command looks hung, so print a
        // status line (the same printInfo pattern the interactive `create-connection` flow uses). Only
        // print when stdout is a real terminal (skip piped/redirected/CI output, where no one is
        // watching and the line would just pollute the stream) and not under `--json` (the stdout line
        // would corrupt the payload). JSON detection is delegated to the CLI's authoritative
        // `ArgumentProvider.isJsonMode()` (wired via `setJsonModeResolver`) so it stays consistent with
        // every other JSON-gated surface.
        if (process.stdout.isTTY && !this._isJsonMode()) {
            printInfo('Opening your browser to sign in. Complete the sign-in there to continue...');
        }
        const result = await this._msalClient.acquireTokenInteractive(request);
        this._tenantId = result.tenantId;
        if (result.account)
            this._ensureActivePointer(result.account);
        return result;
    }
    _getInteractiveLoginRequest(resource, loginHint) {
        const openBrowser = async (url) => {
            try {
                await open(url, { wait: false });
            }
            catch (error) {
                // Auto-open can fail on headless/remote machines or when no default browser is set.
                // Surface the URL (on stderr) so the user can finish sign-in manually — the interactive
                // call keeps waiting for the redirect, so a manual open still completes the flow.
                printError(`Couldn't open a browser automatically (${String(error)}).\n` +
                    `Open this URL to finish signing in:\n${url}`);
            }
        };
        return {
            scopes: [resource],
            openBrowser,
            successTemplate: 'You can close this window now.',
            ...(loginHint && { loginHint }),
        };
    }
}
