// @microsoft/power-apps-cli/dist/Authentication/PacCliAuthenticationProvider.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import * as readline from 'readline';
export class PacCliAuthenticationProvider {
    _tokenCache = {};
    async initAsync(_region) {
        // no-op
    }
    async getAccessTokenForResource(resource) {
        return this._getAccessTokenFromResourceInternal(resource);
    }
    getUserTenantId() {
        return undefined;
    }
    async getCachedAccountAsync() {
        // The PAC CLI host owns identity; there is no local cached account. Returning `undefined` skips
        // the eager telemetry upgrade, so we never trigger a PAC token exchange just to send telemetry —
        // the lazy onFirstAuth path upgrades after the verb's first authenticated request instead.
        return undefined;
    }
    async _getAccessTokenFromResourceInternal(resource) {
        if (this._tokenCache[resource]) {
            return this._tokenCache[resource];
        }
        this._requestToken(resource);
        const token = await this._readStdIn();
        if (!token) {
            throw new Error('No token provided');
        }
        this._tokenCache[resource] = token;
        return token;
    }
    _requestToken(resource) {
        console.log('REQUEST_TOKEN ' + resource);
    }
    async _readStdIn() {
        return new Promise((resolve, reject) => {
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout,
                terminal: false,
            });
            // Wait 10 seconds to get the token before throwing an error
            const timeout = setTimeout(() => {
                rl.close();
                reject(new Error('Token input timeout: No token received within 10 seconds'));
            }, 10000);
            rl.once('line', (line) => {
                clearTimeout(timeout);
                rl.close();
                resolve(line.trim());
            });
        });
    }
}
