// @microsoft/power-apps-cli/dist/Logger/CliLogger.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getEnvironmentByName } from '@microsoft/power-apps-actions';
import { randomUUID } from 'crypto';
import packageJson from '../../package.json' with { type: 'json' };
import { ConsoleWriter } from './ConsoleWriter.js';
import { EUGeoRegions } from './Constants.js';
import { resolveUserTelemetrySettings } from './LoggerSettings.js';
import { OneDSWriter } from './OneDSWriter.js';
/**
 * Utility method to create a Scenario
 *
 * @param {ITelemetryWriter[]} telemetryWriters
 * @param {string} name
 * @param {object} [data]
 * @returns {Scenario}
 */
function createScenarioLogger(telemetryWriters, name, data) {
    const scenarioGuid = randomUUID();
    const _start = Date.now();
    try {
        telemetryWriters.forEach((writer) => writer.startScenario(name, { scenarioGuid, ...data, scenarioStartTime: _start }));
    }
    catch {
        // No-op. Telemetry failures should not block CLI operations
    }
    const base = {
        scenarioId: scenarioGuid,
        complete: (completeData) => {
            try {
                telemetryWriters.forEach((writer) => writer.completeScenario(name, {
                    scenarioGuid,
                    elapsedTime: Date.now() - _start,
                    totalElapsedTime: Date.now() - _start,
                    ...completeData,
                }));
            }
            catch {
                // No-op. Telemetry failures should not block CLI operations
            }
        },
        failure: (failureData) => {
            try {
                telemetryWriters.forEach((writer) => writer.scenarioFailed(name, {
                    scenarioGuid,
                    elapsedTime: Date.now() - _start,
                    totalElapsedTime: Date.now() - _start,
                    ...failureData,
                }));
            }
            catch {
                // No-op. Telemetry failures should not block CLI operations
            }
        },
        completeWithError: (dataWithErr) => {
            try {
                telemetryWriters.forEach((writer) => writer.scenarioCompletedWithError(name, {
                    scenarioGuid,
                    elapsedTime: Date.now() - _start,
                    totalElapsedTime: Date.now() - _start,
                    ...dataWithErr,
                }));
            }
            catch {
                // No-op. Telemetry failures should not block CLI operations
            }
        },
    };
    base._start = _start;
    return base;
}
/**
 * Logger implementation for the CLI.
 *
 * Telemetry is decoupled from authentication so that no verb can ever trigger a telemetry-driven
 * sign-in. When remote telemetry is enabled and a region + environmentId are known, an
 * unauthenticated OneDS writer is pushed eagerly at `create()` time (global 1DS endpoint, no
 * geoLocation); 1DS submission uses its own ingestion credentials, so events flow from event #1
 * without any Entra token. That writer is upgraded to the EU/tenant-aware authed version via one of
 * two paths, neither of which can pop a browser:
 *
 * - `tryEagerUpgradeAsync` (from `ArgumentProvider.runVerb`) — upgrades synchronously only when a
 *   cached account already exists (`getCachedAccountAsync()`, a token-free cache read).
 * - `CliHttpClient.onFirstAuth` (registered here) — upgrades in the background after the verb's
 *   first successful authenticated HTTP call, so the metadata fetch rides a token that already
 *   exists.
 *
 * Verbs that never make an authed call and have no cached account stay on the unauthenticated
 * writer — events still reach 1DS, just on the global endpoint with no geoLocation or tenantId.
 */
export class CliLogger {
    _telemetryWriters = [];
    _unauthWriter;
    // Writers removed from the active set by an upgrade. They no longer receive new events but are
    // still flushed at teardown so their buffered pre-upgrade events aren't dropped on exit.
    _discardedWriters = [];
    _upgradeFired = false;
    _upgradeConfig;
    constructor(telemetryWriters) {
        this._telemetryWriters = telemetryWriters;
    }
    static async create(cliSettings, httpClient, authenticationProvider) {
        const environmentId = cliSettings.appConfig?.environmentId;
        const region = cliSettings.appConfig?.region;
        const appId = cliSettings.appConfig?.appId;
        const sessionId = randomUUID();
        const telemetryWriters = [];
        const userTelemetrySettings = resolveUserTelemetrySettings();
        const addConsoleWriter = () => telemetryWriters.push(new ConsoleWriter());
        if (userTelemetrySettings.consoleOnly) {
            addConsoleWriter();
            return new CliLogger(telemetryWriters);
        }
        if (userTelemetrySettings.outputToConsole) {
            addConsoleWriter();
        }
        const instance = new CliLogger(telemetryWriters);
        if (userTelemetrySettings.enabled && region && environmentId) {
            // Push the unauthenticated writer immediately (global 1DS endpoint, no geoLocation). 1DS uses
            // its own ingestion credentials, so events flow with zero auth coupling; the upgrade paths
            // below swap in the EU/tenant-aware writer once auth is known to work.
            const unauth = new OneDSWriter(sessionId, region, 'global', environmentId, appId ?? '', undefined, packageJson.version, cliSettings.source);
            instance._unauthWriter = unauth;
            instance._telemetryWriters.push(unauth);
            instance._upgradeConfig = { sessionId, region, environmentId, appId, cliSettings };
            // Lazy upgrade: fires once after the verb's first successful authed HTTP call, by which point
            // we know auth works and the metadata fetch can't trigger a browser.
            httpClient.onFirstAuth(() => instance._upgradeToAuthedWriterAsync(authenticationProvider));
        }
        return instance;
    }
    /**
     * Eager upgrade: when a cached account already exists at verb-start time, swap in the authed
     * writer synchronously so even the first telemetry event lands on it. Called from
     * `ArgumentProvider.runVerb` after `authProvider.initAsync(region)`. No-op when not signed in (so
     * we never block startup — or open a browser — on a metadata fetch) or when already upgraded.
     */
    async tryEagerUpgradeAsync(authenticationProvider) {
        if (this._upgradeFired || !this._upgradeConfig) {
            return;
        }
        // Explicit `!== undefined` (not a truthy check) so a provider returning an empty-string account
        // identifier still counts as "signed in" per the `string | undefined` contract.
        if ((await authenticationProvider.getCachedAccountAsync()) !== undefined) {
            await this._upgradeToAuthedWriterAsync(authenticationProvider);
        }
    }
    /**
     * Replaces the unauthenticated OneDS writer with an EU/tenant-aware authed one. Best-effort: on
     * any failure the unauthenticated writer stays in place and events keep reaching the global
     * endpoint. Guarded so the eager and lazy paths can't both upgrade.
     */
    async _upgradeToAuthedWriterAsync(authenticationProvider) {
        if (this._upgradeFired || !this._upgradeConfig || !this._unauthWriter) {
            return;
        }
        this._upgradeFired = true;
        const { sessionId, region, environmentId, appId, cliSettings } = this._upgradeConfig;
        try {
            const metadata = await getEnvironmentByName(environmentId);
            let telemetryLocation = 'global';
            if (metadata?.properties?.location && EUGeoRegions.has(metadata.properties.location)) {
                telemetryLocation = 'eu';
            }
            const authed = new OneDSWriter(sessionId, region, telemetryLocation, environmentId, appId ?? '', metadata?.properties?.cluster?.geoShortName, packageJson.version, cliSettings.source);
            // Swap the unauth writer in place so new events go only to the authed writer, and retain the
            // replaced writer in `_discardedWriters` so the final `flush()` (boot's teardown) still drains
            // its buffered pre-upgrade events (e.g. the `Initialize` event emitted before a mid-session
            // sign-in) rather than dropping them on exit. Those events predate sign-in and carry no
            // tenantId, so the global endpoint is the correct destination.
            const previousUnauth = this._unauthWriter;
            const idx = this._telemetryWriters.indexOf(previousUnauth);
            if (idx >= 0) {
                this._telemetryWriters[idx] = authed;
            }
            else {
                this._telemetryWriters.push(authed);
            }
            this._unauthWriter = authed;
            this._discardedWriters.push(previousUnauth);
            // Stamp tenantId AFTER the swap so it lands on the authed writer (and any ConsoleWriter), not
            // the discarded unauth one. The `getEnvironmentByName` call above already triggered a silent
            // token acquisition, so the provider's tenantId is populated by now.
            const tenantId = authenticationProvider.getUserTenantId();
            if (tenantId) {
                this.setCustomDimension('tenantId', tenantId);
            }
        }
        catch {
            // Best-effort. Unauth writer stays in place; events keep going to the global endpoint.
        }
    }
    setCustomDimension(_key, _value) {
        try {
            if (!_key || !_value) {
                return;
            }
            this._telemetryWriters.forEach((writer) => {
                writer.setCustomDimension(_key, _value);
            });
        }
        catch {
            // No-op. Telemetry failures should not block CLI operations
        }
    }
    trackActivityEvent(eventName, eventData) {
        try {
            if (this._telemetryWriters.length === 0) {
                return;
            }
            this._telemetryWriters.forEach((writer) => writer.trackActivityEvent(eventName, eventData));
        }
        catch {
            // No-op. Telemetry failures should not block CLI operations
        }
    }
    trackErrorEvent(eventName, eventData) {
        try {
            if (this._telemetryWriters.length === 0) {
                return;
            }
            this._telemetryWriters.forEach((writer) => writer.trackErrorEvent(eventName, eventData));
        }
        catch {
            // No-op. Telemetry failures should not block CLI operations
        }
    }
    trackScenario(eventName, eventData) {
        try {
            if (this._telemetryWriters.length === 0) {
                // Return a no-op scenario so callers don't have to check telemetry state
                return {
                    scenarioId: 'disabled',
                    complete: () => undefined,
                    failure: () => undefined,
                    completeWithError: () => undefined,
                };
            }
            return createScenarioLogger(this._telemetryWriters, eventName, eventData);
        }
        catch {
            // No-op. Telemetry failures should not block CLI operations
            return {
                scenarioId: 'error',
                complete: () => undefined,
                failure: () => undefined,
                completeWithError: () => undefined,
            };
        }
    }
    stringifyError(err) {
        if (!err) {
            return undefined;
        }
        if (err instanceof Error) {
            return `${err.name}: ${err.message}${err.stack ? '\n' + err.stack : ''}`;
        }
        try {
            return JSON.stringify(err);
        }
        catch {
            return String(err);
        }
    }
    flush() {
        // Flush active writers plus any writers an upgrade discarded, so their buffered events are
        // drained even though they no longer receive new events.
        return Promise.all([...this._telemetryWriters, ...this._discardedWriters].map((writer) => writer.flush())).then(() => undefined);
    }
}
