// @microsoft/power-apps-actions/dist/services/connectivity/ConnectivityService.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getPlayerServiceConfig } from '../initializePlayerServices.js';
import { getPowerPlatformApiEnvironmentConfig } from '../PowerPlatformApiUtils.js';
export async function getConnectorAsync(connectorId, logger) {
    const scenario = logger?.trackScenario('GetConnectorAsync', {
        connectorId,
    });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const url = `https://${endpoint}/connectivity/connectors/${connectorId}?api-version=1&$filter=environment+eq+'${environmentName}'`;
        const response = await httpClient.get(url, {
            authResource,
        });
        additionalScenarioData = { url, authResource, status: response.status };
        if (response.status === 200) {
            scenario?.complete(additionalScenarioData);
        }
        else {
            scenario?.completeWithError(additionalScenarioData);
        }
        return response.data;
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
export async function getConnectionByNameAsync(connectionName, connectorId, logger) {
    const scenario = logger?.trackScenario('GetConnectionByNameAsync', {
        connectionName,
        connectorId,
    });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const url = `https://${endpoint}/connectivity/connectors/${connectorId}/connections/${connectionName}?api-version=1&$filter=environment+eq+'${environmentName}'`;
        const response = await httpClient.get(url, {
            authResource,
        });
        additionalScenarioData = { url, authResource, status: response.status };
        if (response.status === 200) {
            scenario?.complete(additionalScenarioData);
        }
        else {
            scenario?.completeWithError(additionalScenarioData);
        }
        return response.data;
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
export async function listConnectionsForConnectorAsync(connectorId, logger) {
    const scenario = logger?.trackScenario('ListConnectionsForConnectorAsync', { connectorId });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const url = `https://${endpoint}/connectivity/connectors/${connectorId}/connections?api-version=1&$filter=environment+eq+'${environmentName}'`;
        const response = await httpClient.get(url, { authResource });
        additionalScenarioData = { url, authResource, status: response.status };
        if (response.status === 200) {
            scenario?.complete(additionalScenarioData);
        }
        else {
            scenario?.completeWithError(additionalScenarioData);
        }
        return response.data.value ?? [];
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
export async function listAllConnectionsAsync(search, logger) {
    const scenario = logger?.trackScenario('ListAllConnectionsAsync', { search });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const initialUrl = `https://${endpoint}/connectivity/connections?api-version=1&$filter=environment+eq+'${environmentName}'&%24expand=api`;
        additionalScenarioData = { url: initialUrl, authResource };
        const allConnections = [];
        let nextUrl = initialUrl;
        while (nextUrl) {
            const currentUrl = nextUrl;
            const response = await httpClient.get(currentUrl, {
                authResource,
            });
            additionalScenarioData = { url: currentUrl, authResource, status: response.status };
            if (response.status !== 200) {
                scenario?.completeWithError(additionalScenarioData);
                throw new Error(`Failed to list connections. Received status ${response.status}.`);
            }
            allConnections.push(...(response.data.value ?? []));
            nextUrl = response.data['@odata.nextLink'];
        }
        let connections = allConnections;
        if (search) {
            const lower = search.toLowerCase();
            connections = connections.filter((c) => c.properties.displayName.toLowerCase().includes(lower) ||
                c.properties.apiId.toLowerCase().includes(lower));
        }
        scenario?.complete({ ...additionalScenarioData, count: connections.length });
        return connections;
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
export function isConnectionShareable(connector, authType) {
    if (authType) {
        const paramSet = connector.properties.connectionParameterSets?.values?.find((x) => x.name === authType);
        if (paramSet) {
            return paramSet.metadata?.allowSharing ?? false;
        }
    }
    return connector.properties.metadata?.allowSharing ?? false;
}
/**
 * Creates a new connection via PUT to the connectivity API.
 *
 * When `ssoSetName` is provided (from `validateSsoEligible`), the body uses
 * `connectionParametersSet: { name, values: {} }` — required for connectors that
 * use `connectionParameterSets` (e.g. SQL). Otherwise falls back to
 * `connectionParameters: {}` for flat-parameter connectors (e.g. Office 365).
 */
export async function putConnectionAsync(connectorId, connectionId, displayName, logger, ssoSetName) {
    const scenario = logger?.trackScenario('PutConnectionAsync', { connectorId, connectionId });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const url = `https://${endpoint}/connectivity/connectors/${connectorId}/connections/${connectionId}?api-version=1&$expand=ConsentLink`;
        const connectionBody = ssoSetName
            ? { connectionParametersSet: { name: ssoSetName, values: {} } }
            : { connectionParameters: {} };
        const body = {
            properties: {
                displayName,
                ...connectionBody,
                environment: { name: environmentName },
                consentInfo: { redirectUrl: 'https://www.microsoft.com/' },
            },
        };
        additionalScenarioData = { url, authResource };
        const response = await httpClient.put(url, { authResource, body });
        additionalScenarioData.status = response.status;
        if (response.status !== 200 && response.status !== 201) {
            scenario?.completeWithError(additionalScenarioData);
            throw new Error(`Failed to create connection for connector '${connectorId}': HTTP ${response.status}`);
        }
        scenario?.complete(additionalScenarioData);
        return response.data;
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
/**
 * Deletes a connection. Used to clean up orphaned connections when a
 * later step of the creation flow fails.
 */
export async function deleteConnectionAsync(connectorId, connectionId, logger) {
    const scenario = logger?.trackScenario('DeleteConnectionAsync', { connectorId, connectionId });
    let additionalScenarioData = {};
    try {
        const { endpoint, authResource } = getPowerPlatformApiEnvironmentConfig();
        const { httpClient, environmentName } = getPlayerServiceConfig();
        const url = `https://${endpoint}/connectivity/connectors/${connectorId}/connections/${connectionId}` +
            `?api-version=1&$filter=environment+eq+'${environmentName}'`;
        additionalScenarioData = { url, authResource };
        const response = await httpClient.delete(url, { authResource });
        scenario?.complete({ ...additionalScenarioData, status: response.status });
    }
    catch (error) {
        scenario?.failure({ error, ...additionalScenarioData });
        throw error;
    }
}
/**
 * Returns true if the given parameter map is SSO-eligible:
 *   1. Has at least one `type === 'oauthSetting'` entry.
 *   2. That oauthSetting has `oAuthSettings.properties.IsOnbehalfofLoginSupported === true`
 *      (connector must support the OBO silent token flow).
 *   3. All other entries have `uiDefinition.constraints.hidden === 'true'`
 *      and `uiDefinition.constraints.required === 'false'`.
 * An empty/absent map is considered eligible (no user input needed).
 */
function isParamSetSsoEligible(params) {
    if (!params || Object.keys(params).length === 0)
        return true;
    const oauthParams = Object.values(params).filter((p) => p.type === 'oauthSetting');
    if (oauthParams.length === 0)
        return false;
    if (!oauthParams.some((p) => p.oAuthSettings?.properties?.IsOnbehalfofLoginSupported === true)) {
        return false;
    }
    return !Object.entries(params).some(([, p]) => {
        if (p.type === 'oauthSetting')
            return false;
        const c = p.uiDefinition?.constraints;
        return c?.hidden !== 'true' || c?.required !== 'false';
    });
}
/**
 * Validates that a connector is eligible for silent SSO connection creation.
 *
 * When `connectionParameterSets` is present, it takes precedence over
 * `connectionParameters`. The function returns the name of the first
 * SSO-eligible set, or throws if none qualifies.
 *
 * When only `connectionParameters` is present, the same two-condition check
 * applies. Returns `null` to signal "use flat connectionParameters".
 *
 * @returns The name of the SSO-eligible `connectionParameterSet` to use, or
 *          `null` if the connector uses flat `connectionParameters`.
 */
export function validateSsoEligible(connector, connectorId) {
    const notEligibleError = `Connector '${connectorId}' is not eligible for silent SSO.`;
    const sets = connector.properties.connectionParameterSets?.values;
    if (sets && sets.length > 0) {
        // connectionParameterSets takes precedence over connectionParameters
        const ssoSet = sets.find((set) => isParamSetSsoEligible(set.parameters));
        if (!ssoSet) {
            throw new Error(notEligibleError);
        }
        return ssoSet.name;
    }
    // Fall back to flat connectionParameters
    const params = connector.properties.connectionParameters;
    if (!params || Object.keys(params).length === 0)
        return null;
    if (!isParamSetSsoEligible(params)) {
        throw new Error(notEligibleError);
    }
    return null;
}
/**
 * Checks whether a connector supports SSO as its *only* authentication type.
 * Returns `true` when the connector has a single auth type (or flat params)
 * that is SSO-eligible. Returns `false` if there are multiple auth types
 * (even if one is SSO-capable) — in that case the user should choose
 * interactively.
 */
export function isSsoOnlyConnector(connector) {
    const sets = connector.properties.connectionParameterSets?.values;
    if (sets && sets.length > 0) {
        // Multiple auth types → user must choose interactively
        if (sets.length > 1)
            return false;
        return isParamSetSsoEligible(sets[0].parameters);
    }
    // Fall back to flat connectionParameters
    const params = connector.properties.connectionParameters;
    if (!params || Object.keys(params).length === 0)
        return true;
    return isParamSetSsoEligible(params);
}
